File: tests/lua/event_declared.lua
-- event_declared.lua — the runtime half of labelle.event (one DSL, two
-- consumers): the SAME line the declare runner reads as an event-schema
-- declaration evaluates HERE to the event-name string itself, so one
-- binding drives both legs of the bus — labelle.emit(HungerFeed, ...)
-- toward the host and labelle.on(HungerFeed, ...) back from it.
-- labelle.id is plain 0 at runtime. Chunk-scope findings (returned name,
-- id value, the name-validation errors) assert via error(), which evicts
-- the script — so the Fed component existing at all is half the proof.

local HungerFeed = labelle.event("hunger__feed", { entity = labelle.id, amount = 0.5 })

if HungerFeed ~= "hunger__feed" then
    error("event did not return its name")
end
if labelle.id ~= 0 then
    error("labelle.id is not 0 at runtime")
end

local ok, err = pcall(labelle.event, "")
if ok or not tostring(err):find("non%-empty event name") then
    error("empty event name accepted or wrong error: " .. tostring(err))
end
ok, err = pcall(labelle.event, 42)
if ok or not tostring(err):find("non%-empty event name") then
    error("non-string event name accepted or wrong error: " .. tostring(err))
end

local state
local sent = false

labelle.on(HungerFeed, function(ev)
    local s = state:get("Fed")
    s.count = s.count + 1
    s.amount = ev.amount
    state:set("Fed", s)
end)

function init()
    state = Entity.new()
    state:set("Fed", { count = 0, ok = true })
end

function update(dt)
    -- Emit through the binding once, toward the host (the mock captures
    -- script→host emissions verbatim).
    if sent then return end
    sent = true
    labelle.emit(HungerFeed, { entity = labelle.u64str(state.id), amount = 0.5 })
end

File: tests/ruby/event_declared.rb
# event_declared.rb — the runtime half of Labelle.event (one DSL, two
# consumers): the SAME line the declare runner reads as an event-schema
# declaration evaluates HERE to the frozen event-name string, so one
# constant drives both legs of the bus — Labelle.emit(HungerFeed, ...)
# toward the host and Labelle.on(HungerFeed) back from it. Labelle.id is
# plain 0 at runtime. Chunk-scope findings (frozen return, id value, the
# name-validation raises) fold into `ok`, reported through the Fed
# component — a false there names the failing leg via the raise below.

HungerFeed = Labelle.event "hunger__feed", entity: Labelle.id, amount: 0.5

raise "event did not return its name" unless HungerFeed == "hunger__feed"
raise "event name not frozen" unless HungerFeed.frozen?
raise "Labelle.id is not 0 at runtime" unless Labelle.id == 0

begin
  Labelle.event ""
  raise "empty event name accepted"
rescue ArgumentError => e
  raise "wrong empty-name error" unless e.message.include?("non-empty event name")
end

begin
  Labelle.event 42
  raise "non-string event name accepted"
rescue ArgumentError => e
  raise "wrong non-string error" unless e.message.include?("non-empty event name")
end

def init
  @state = Labelle::Entity.create
  @state.set("Fed", count: 0, ok: true)

  Labelle.on(HungerFeed) do |ev|
    s = @state.get("Fed")
    s[:count] += 1
    s[:amount] = ev[:amount]
    @state.set("Fed", s)
  end
end

def update(dt)
  # Emit through the constant once, toward the host (the mock captures
  # script→host emissions verbatim).
  return if @sent
  @sent = true
  Labelle.emit(HungerFeed, entity: Labelle.u64str(@state.id), amount: 0.5)
end
